### PR TITLE
feat: support tickets history

### PR DESCRIPTION
**Pull request by ytvnr**
Fixes gravitee-io/issues#4435

### CODE REVIEW

File: gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/param/TicketsParam.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    @ApiParam(value = "The api identifier used to filter tickets")
+    private String api;
+
+    @QueryParam("field")
+    @ApiParam(value = "The field to query when doing `sort` queries")
+    private String field;
+    
+    @QueryParam("order")
+    @ApiParam(value = "ASC or DESC to define the sort order")
+    private String order;
```
Comment (phiz71) [inline gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/param/TicketsParam.java:39]: See next comment in openapi.yaml

File: gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/resources/openapi.yaml [hunk 2]
```diff
@@ -2015,6 +2015,28 @@ paths:
...
+      parameters:
+        - $ref: '#/components/parameters/apiIdQueryParam'
+        - $ref: '#/components/parameters/pageNumberParam'
+        - $ref: '#/components/parameters/pageSizeParam'
+        - $ref: '#/components/parameters/logOrderQueryParam'
+        - $ref: '#/components/parameters/fieldQueryParam'
+      description: |
+        List all tickets written by current user
+
+      operationId: search
```
Comment (phiz71) [inline gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/resources/openapi.yaml:2030]: getTickets

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java [hunk 3]
```diff
@@ -59,70 +73,167 @@
...
+                    new PageableBuilder()
+                            .pageNumber(pageable.getPageNumber() - 1)
+                            .pageSize(pageable.getPageSize())
+                            .build()
+            );
+
+            List<TicketEntity> entities = tickets.getContent()
+                    .stream()
+                    .map(t -> {
+                        final Ticket ticket = new Ticket(t);
```
Comment (phiz71) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java:175]: why creating a new ticket ?

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java [hunk 4]
```diff
@@ -59,70 +73,167 @@
...
         }
+    }
+
+    @Override
+    public TicketEntity findById(String ticketId) {
+        try {
+            return ticketRepository
+                    .findById(ticketId)
+                    .map(t -> {
+                        Ticket ticket = new Ticket(t);
```
Comment (phiz71) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java:214]: why creating a new ticket ?

**Review by phiz71 — CHANGES_REQUESTED** [review]

File: gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/PlatformTicketsResource.java [hunk 5]
```diff
@@ -52,4 +69,37 @@ public Response create(@Valid @NotNull final NewTicketEntity ticketEntity) {
...
         return Response.created(URI.create("")).build();
     }
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = "Search for platform tickets written by current user")
+    @ApiResponses({
+            @ApiResponse(code = 200, message = "List platform tickets written by current user", response = TicketEntity.class, responseContainer = "Page"),
+            @ApiResponse(code = 500, message = "Internal server error")})
+    public Page<TicketEntity> search(
```
[the remark this answers is not in the document]
Comment (ytvnr) [reply gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/PlatformTicketsResource.java:79]: resolved

File: gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/PlatformTicketsResource.java [hunk 6]
```diff
@@ -52,4 +69,37 @@ public Response create(@Valid @NotNull final NewTicketEntity ticketEntity) {
...
+    @GET
+    @Path("/{ticket}")
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = "Get a specific ticket")
+    @ApiResponses({
+            @ApiResponse(code = 200, message = "Get a platform ticket", response = TicketEntity.class),
+            @ApiResponse(code = 500, message = "Internal server error")})
+    public Response findTicket(@PathParam("ticket") String ticketId) {
+
+        TicketEntity ticketEntity = ticketService.findById(ticketId);
```
[the remark this answers is not in the document]
Comment (ytvnr) [reply gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/PlatformTicketsResource.java:101]: i replaced by getTicket

File: gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/TicketsResource.java [hunk 7]
```diff
@@ -47,4 +65,31 @@ public Response create(@Valid @NotNull(message = "Input must not be null.") fina
         ticketService.create(getAuthenticatedUser(), ticketMapper.convert(ticketInput));
         return Response.created(URI.create("")).build();
     }
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response search(
```
[the remark this answers is not in the document]
Comment (ytvnr) [reply gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/TicketsResource.java:71]: resolved

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java (hunk 3 above)
Comment (ytvnr) [reply gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java:175]: indeed, no need.
resolved

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java (hunk 4 above)
Comment (ytvnr) [reply gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java:214]: resolved !

File: gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/param/TicketsParam.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+import io.swagger.annotations.ApiParam;
+
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.QueryParam;
+
+/**
+ * @author Yann TAVERNIER (yann.tavernier at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class TicketsParam {
```
[the remark this answers is not in the document]
Comment (ytvnr) [reply gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/param/TicketsParam.java:27]: resolved

File: gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/PlatformTicketsResource.java [hunk 9]
```diff
@@ -15,8 +15,17 @@
  */
 package io.gravitee.rest.api.management.rest.resource;
 
+import io.gravitee.common.data.domain.Page;
 import io.gravitee.common.http.MediaType;
+import io.gravitee.repository.analytics.query.Order;
+import io.gravitee.repository.analytics.query.SortBuilder;
```
[the remark this answers is not in the document]
Comment (ytvnr) [reply gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/PlatformTicketsResource.java:21]: resolved

File: gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/param/TicketsParam.java (hunk 1 above)
Comment (ytvnr) [reply gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/param/TicketsParam.java:39]: done

File: gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/resources/openapi.yaml (hunk 2 above)
Comment (ytvnr) [reply gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/resources/openapi.yaml:2030]: done

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java [hunk 10]
```diff
@@ -59,70 +78,138 @@
...
-                        .build());
-        sendUserNotification(user, api, applicationEntity);
+    @Override
+    public TicketEntity findById(String ticketId) {
+        try {
+            return ticketRepository
+                    .findById(ticketId)
+                    .map(this::getApiNameAndApplicationName)
+                    .map(this::convert)
+                    .orElse(null);
```
Comment (phiz71) [inline gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java:207]: should throw an exception instead

File: gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/PlatformTicketsResource.java [hunk 11]
```diff
@@ -52,4 +64,45 @@ public Response createPlatformTicket(@Valid @NotNull final NewTicketEntity ticke
...
+                pageable.toPageable());
+    }
+
+    @GET
+    @Path("/{ticket}")
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = "Get a specific ticket")
+    @ApiResponses({
+            @ApiResponse(code = 200, message = "Get a platform ticket", response = TicketEntity.class),
+            @ApiResponse(code = 500, message = "Internal server error")})
```
Comment (phiz71) [inline gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/PlatformTicketsResource.java:101]: "404" is missing in the documentation

**Review by phiz71 — CHANGES_REQUESTED** [review]

File: gravitee-rest-api-model/src/main/java/io/gravitee/rest/api/model/common/Sortable.java [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.rest.api.model.common;
+
+public interface Sortable {
+    String getField();
+
+    boolean getOrder();
```
[the remark this answers is not in the document]
Comment (ytvnr) [reply gravitee-rest-api-model/src/main/java/io/gravitee/rest/api/model/common/Sortable.java]: resolved

File: gravitee-rest-api-model/src/main/java/io/gravitee/rest/api/model/common/SortableImpl.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        this.order = order;
+    }
+
+    @Override
+    public String getField() {
+        return field;
+    }
+
+    @Override
+    public boolean getOrder() {
```
[the remark this answers is not in the document]
Comment (ytvnr) [reply gravitee-rest-api-model/src/main/java/io/gravitee/rest/api/model/common/SortableImpl.java]: resolved

File: gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/param/TicketsParam.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    @ApiParam(value = "The api identifier used to filter tickets")
+    private String api;
+
+    @QueryParam("applicationId")
+    @ApiParam(value = "The application identifier used to filter tickets")
+    private String application;
+
+    @QueryParam("field")
+    @ApiParam(value = "The field to query when doing `sort` queries")
+    private String field;
```
[the remark this answers is not in the document]
Comment (ytvnr) [reply gravitee-rest-api-portal/gravitee-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/param/TicketsParam.java]: resolved

File: gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java (hunk 10 above)
Comment (ytvnr) [reply gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/TicketServiceImpl.java]: resolved

File: gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/PlatformTicketsResource.java (hunk 11 above)
Comment (ytvnr) [reply gravitee-rest-api-management/gravitee-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/PlatformTicketsResource.java:102]: resolved

**Review by phiz71 — APPROVED** [review]

**Pull request merged**